### PR TITLE
Name resolution delay

### PR DESCRIPTION
**Pull request by AgraVator**
This finishes the remaining work in https://github.com/grpc/grpc-java/pull/12807

### CODE REVIEW

File: core/src/test/java/io/grpc/internal/ForwardingClientStreamTracerTest.java [hunk 1]
```diff
@@ -40,6 +40,19 @@ public void allMethodsForwarded() throws Exception {
         Collections.<Method>emptyList());
   }
 
+  @Test
+  public void attemptDelayMethodsForwarded() {
+    TestClientStreamTracer tracer = new TestClientStreamTracer();
+    tracer.recordAttemptDelayStart("connecting", "test");
+    org.mockito.Mockito.verify(mockDelegate).recordAttemptDelayStart("connecting", "test");
```
Comment (kannanjgithub) [inline core/src/test/java/io/grpc/internal/ForwardingClientStreamTracerTest.java]: nit: static import verify.

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryMetricsModule.java [hunk 2]
```diff
@@ -387,6 +377,11 @@ static final class CallAttemptsTracerFactory extends ClientStreamTracer.Factory
     private final List<OpenTelemetryPlugin.ClientCallPlugin> callPlugins;
     private final Context otelContext;
     private Status status;
+    @GuardedBy("this")
+    @Nullable private Stopwatch activeCallDelayStopwatch;
+    @GuardedBy("this")
+    @Nullable private String activeCallDelayType;
+    private final io.opentelemetry.api.common.Attributes callLevelBaseAttributes;
```
Comment (kannanjgithub) [inline opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryMetricsModule.java]: nit: Use the imported Attributes.

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryMetricsModule.java [hunk 3]
```diff
@@ -420,6 +415,7 @@ static final class CallAttemptsTracerFactory extends ClientStreamTracer.Factory
             CUSTOM_LABEL_KEY, callOptions.getOption(Grpc.CALL_OPTION_CUSTOM_LABEL));
       }
       io.opentelemetry.api.common.Attributes attribute = builder.build();
+      this.callLevelBaseAttributes = attribute;
```
Comment (kannanjgithub) [inline opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryMetricsModule.java]: nit: Why not directly assign the built object without using the local variable `attribute`?

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java [hunk 4]
```diff
@@ -185,25 +197,78 @@ void callEnded(io.grpc.Status status) {
...
+    }
+
+    @Override
+    public synchronized void recordCallDelayReasonChanged(String delayReason) {
+      if (!GrpcOpenTelemetry.isDelayObservabilityEnabled()
+          || isCallEnded()
+          || activeCallDelaySpan == null) {
+        return;
+      }
+      String type = activeCallDelayType;
```
Comment (kannanjgithub) [inline opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java]: nit: No need for this local variable as the method itself is syncrhonized. Same for local variable `delaySpan` in `recordCallDelayEnd` below.

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java [hunk 5]
```diff
@@ -185,25 +197,78 @@ void callEnded(io.grpc.Status status) {
...
+      activeCallDelayType = delayType;
+      Span delaySpan = otelTracer.spanBuilder("Call Delay")
+          .setParent(Context.current().with(clientSpan))
+          .setAttribute("grpc.delay_type", delayType)
+          .startSpan();
+      activeCallDelaySpan = delaySpan;
+      delaySpan.addEvent(
+          "Delay state transition",
+          Attributes.of(
+              AttributeKey.stringKey("grpc.delay_type"), delayType,
```
Comment (kannanjgithub) [inline opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java]: Shouldn't we have the null check for `delayType` like in the method below?

File: opentelemetry/src/test/java/io/grpc/opentelemetry/GrpcOpenTelemetryTest.java [hunk 6]
```diff
@@ -179,6 +239,242 @@ public void configureChannelBuilder_registersMetricSink() {
...
+      public void onClose(Status status, Metadata trailers) {
+        latch.countDown();
+      }
+    }, new Metadata());
+    call.sendMessage("request_payload");
+    call.halfClose();
+    call.request(1);
+    assertThat(latch.await(5, TimeUnit.SECONDS)).isTrue();
+
+    io.opentelemetry.sdk.testing.assertj.OpenTelemetryAssertions
```
Comment (kannanjgithub) [inline opentelemetry/src/test/java/io/grpc/opentelemetry/GrpcOpenTelemetryTest.java]: nit: Import and use.

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryMetricsModule.java (hunk 3 above)
Comment (AgraVator) [reply opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryMetricsModule.java]: Was pre existing, updated

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryMetricsModule.java (hunk 2 above)
Comment (AgraVator) [reply opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryMetricsModule.java]: updated

**Comment by kannanjgithub** [comment]
Can you check the missing code coverages from the report
```
./gradlew :grpc-api:test :grpc-api:jacocoTestReport :grpc-core:test :grpc-core:jacocoTestReport :grpc-opentelemetry:test :grpc-opentelemetry:jacocoTestReport
```
and fix them?

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java [hunk 7]
```diff
@@ -428,11 +430,11 @@ private PendingStream(PickSubchannelArgs args, ClientStreamTracer[] tracers,
      * structured transition event is appended to the active span without span re-creation.
      */
     synchronized void updateDelay(@Nullable String newType, @Nullable String newReason) {
-      if (getRealStream() != null) {
+      if (getRealStream() != null || delayEnded) {
```
Comment (kannanjgithub) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java:433]: Branch getRealStream() == null && delayEnded == true is missing coverage. To cover it, a unit test only needs to invoke `endDelay()` directly on a pending stream while `getRealStream() == null` and then call `updateDelay(...)`:


```
// Example in DelayedClientTransportTest.java
PendingStream pendingStream = ...;
pendingStream.endDelay(); // sets delayEnded = true while getRealStream() is still null
pendingStream.updateDelay("connecting", "new reason"); // triggers missing branch > returns
```

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java [hunk 8]
```diff
@@ -428,11 +430,11 @@ private PendingStream(PickSubchannelArgs args, ClientStreamTracer[] tracers,
...
         return;
       }
+      // Delay type changed (e.g., from RLS lookup to connecting). End the previous delay.
       if (!Objects.equals(activeDelayType, newType)) {
-        // Delay type changed (e.g., from RLS lookup to connecting). End the previous delay.
         if (activeDelayType != null) {
```
Comment (kannanjgithub) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java:438]: The is reported missing coverage for the branch when `activeDelayType` is null but it cannot be meaningfully tested because `determineQueuingDelayType` always returns non null value and there are other complications if we make this unconditional.

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryMetricsModule.java [hunk 9]
```diff
@@ -579,6 +574,40 @@ void recordFinishedCall(CallOptions callOptions) {
         );
       }
     }
+
+    @Override
+    public synchronized void recordCallDelayStart(String delayType, String delayReason) {
+      if (!GrpcOpenTelemetry.isDelayObservabilityEnabled()
+          || (activeCallDelayStopwatch != null && Objects.equals(activeCallDelayType, delayType))) {
```
Comment (kannanjgithub) [inline opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryMetricsModule.java:581]: When `activeCallDelayStopwatch != null` is true, invoking `recordCallDelayStart` with a different `delayType` while an existing call delay is already active to transition between delay segments is not exercised.

 We can cover this branch by adding a transition call with a different delay type (e.g. `"connecting"`) to `clientCallDelayDuration_duplicateStart_ignoresSecondStart` in `OpenTelemetryMetricsModuleTest.java`:

 ```java
   @Test
   public void clientCallDelayDuration_duplicateStart_ignoresSecondStart() {
     OpenTelemetryMetricsResource resource = GrpcOpenTelemetry.createMetricInstruments(
         openTelemetryTesting.getOpenTelemetry().getMeterProvider().get("grpc-java"),
         ImmutableMap.of("grpc.client.call.delay.duration", true),
         false);
     OpenTelemetryMetricsModule module = new OpenTelemetryMetricsModule(
         new FakeClock().getStopwatchSupplier(), resource, emptyList(), emptyList());
     OpenTelemetryMetricsModule.CallAttemptsTracerFactory factory =
         new OpenTelemetryMetricsModule.CallAttemptsTracerFactory(
             module, "target:///", CallOptions.DEFAULT, method.getFullMethodName(),
             emptyList(), io.opentelemetry.context.Context.root());
 
     factory.recordCallDelayStart("resolving", "first start");
     // Duplicate call with same delay type is ignored
     factory.recordCallDelayStart("resolving", "second start");
 
     // Transition to a different delay type while active (exercises Objects.equals == false)
     factory.recordCallDelayStart("connecting", "transition to connecting");
     factory.recordCallDelayEnd();
 
     assertThat(openTelemetryTesting.getMetrics())
         .anySatisfy(
             metric - assertThat(metric)
                 .hasName("grpc.client.call.delay.duration"));
   }
 ```

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryMetricsModule.java [hunk 10]
```diff
@@ -579,6 +574,40 @@ void recordFinishedCall(CallOptions callOptions) {
...
+
+    @Override
+    public synchronized void recordCallDelayReasonChanged(String delayReason) {
+    }
+
+    @Override
+    public synchronized void recordCallDelayEnd() {
+      Stopwatch delayStopwatch = activeCallDelayStopwatch;
+      String delayType = activeCallDelayType;
+      if (delayStopwatch != null && delayType != null) {
```
Comment (kannanjgithub) [inline opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryMetricsModule.java:597]: In the test for the null delay type branch, assert that no metric was recorded for null (e.g. assertThat(openTelemetryTesting.getMetrics()).extracting("name").doesNotContain("grpc.client.call.delay.duration") or verify that only points with valid non-null grpc.delay_type attributes exist).

**Review by kannanjgithub — CHANGES_REQUESTED** [review]
There are still some missing branches in codecov report. I have commented about missing coverages for OpenTelemetryMetricsModule. 
Also address still existing coverages in the report except for cases where it cannot be done without changing class visibility or other reasons.

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java (hunk 7 above)
Comment (kannanjgithub) [reply core/src/main/java/io/grpc/internal/DelayedClientTransport.java:433]: Lets not do this, I didn't realize this change requires changing the class access modifier for PendingStream.

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java [hunk 11]
```diff
@@ -1015,25 +1018,27 @@ private final class PendingCall<ReqT, RespT> extends DelayedClientCall<ReqT, Res
...
-        }
-        if (shouldStart) {
-          for (ClientStreamTracer.Factory factory : callOptions.getStreamTracerFactories()) {
-            factory.recordCallDelayStart(
-                "resolving", "waiting for name resolution or service config");
+            for (ClientStreamTracer.Factory factory : callOptions.getStreamTracerFactories()) {
+              if (callCancelled || delayEnded) {
+                break;
+              }
+              factory.recordCallDelayStart(
```
Comment (kannanjgithub) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java]: Why move this inside `synchronized` since `recordCallDelayStart` is thread-safe, looking at the implementations?

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryMetricsModule.java (hunk 2 above)
Comment (kannanjgithub) [reply opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryMetricsModule.java]: This is unchanged, we don't need to use the fully qualified class name here.

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryMetricsModule.java (hunk 3 above)
Comment (kannanjgithub) [reply opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryMetricsModule.java]: This is still unchanged.

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java [hunk 12]
```diff
@@ -185,8 +196,72 @@ void callEnded(io.grpc.Status status) {
...
+                AttributeKey.stringKey("grpc.delay_type"), delayType,
+                AttributeKey.stringKey("grpc.delay_reason"), delayReason));
+      }
+    }
+
+    @Override
+    public void recordCallDelayReasonChanged(String delayReason) {
+      if (!GrpcOpenTelemetry.isDelayObservabilityEnabled()
+          || isCallEnded()
+          || activeCallDelaySpan == null) {
```
Comment (kannanjgithub) [inline opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java]: Remove these 2.

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java [hunk 13]
```diff
@@ -222,48 +300,63 @@ public void recordAttemptDelayStart(String delayType, String delayReason) {
...
-      delaySpan.addEvent(
-          "Delay state transition",
-          io.opentelemetry.api.common.Attributes.of(
-              AttributeKey.stringKey("grpc.delay_type"), delayType,
-              AttributeKey.stringKey("grpc.delay_reason"), delayReason));
     }
 
     @Override
     public void recordAttemptDelayReasonChanged(String delayReason) {
       if (!GrpcOpenTelemetry.isDelayObservabilityEnabled() || activeDelaySpan == null) {
```
Comment (kannanjgithub) [inline opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java]: Remove `activeDelayspan == null` condition from here.

**Review by kannanjgithub — CHANGES_REQUESTED** [review]

File: api/src/main/java/io/grpc/ClientStreamTracer.java [hunk 14]
```diff
@@ -67,7 +67,7 @@ public void createPendingStream() {
    *
    * @param delayType canonical low-cardinality label categorizing the delay (e.g., "connecting")
    * @param delayReason high-cardinality diagnostic string describing granular runtime conditions
-   * @since 1.82.0
+   * @since 1.84.0
```
Comment (kannanjgithub) [inline api/src/main/java/io/grpc/ClientStreamTracer.java]: Change to 1.85.0.

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java (hunk 5 above)
Comment (AgraVator) [reply opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java]: `  static <T> Attributes of(AttributeKey<T> key, T value) {
    if (key == null || key.getKey().isEmpty() || value == null) {
      return empty();
    }
    return new ArrayBackedAttributes(new Object[] {key, value});
  }
`
will remove the unnecessary checks

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java [hunk 15]
```diff
@@ -1001,8 +1001,6 @@ private final class PendingCall<ReqT, RespT> extends DelayedClientCall<ReqT, Res
       @GuardedBy("this")
       private boolean queuedForResolution;
       @GuardedBy("this")
-      private boolean callCancelled;
-      @GuardedBy("this")
       private boolean delayEnded;
```
Comment (kannanjgithub) [inline core/src/main/java/io/grpc/internal/ManagedChannelImpl.java:1004]: Why remove @GuardedBy("this")?

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java [hunk 16]
```diff
@@ -215,48 +221,39 @@ public void recordCallDelayStart(String delayType, String delayReason) {
...
+            Attributes.of(AttributeKey.stringKey("grpc.delay_reason"), delayReason));
       }
     }
 
     @Override
     public void recordCallDelayReasonChanged(String delayReason) {
-      if (!GrpcOpenTelemetry.isDelayObservabilityEnabled()
-          || isCallEnded()
-          || activeCallDelaySpan == null) {
+      if (!delayObservabilityEnabled || isCallEnded()) {
```
Comment (kannanjgithub) [inline opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java:236]: Remove `isCallEnded()`.

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java [hunk 17]
```diff
@@ -185,25 +200,86 @@ void callEnded(io.grpc.Status status) {
...
       }
+      if (delayObservabilityEnabled) {
+        recordCallDelayEnd();
+      }
       endSpanWithStatus(clientSpan, status);
     }
+
+    @Override
+    public void recordCallDelayStart(String delayType, String delayReason) {
+      if (!delayObservabilityEnabled || isCallEnded()) {
```
Comment (kannanjgithub) [inline opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java:211]: Remove `isCallEnded()`.

**Review by kannanjgithub — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/grpc/internal/ManagedChannelImpl.java (hunk 15 above)
Comment (AgraVator) [reply core/src/main/java/io/grpc/internal/ManagedChannelImpl.java:1004]: Removes callCancelled

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java (hunk 16 above)
Comment (AgraVator) [reply opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java:236]: Will keep it here and remove the one inside the sync block ?

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java (hunk 17 above)
Comment (AgraVator) [reply opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java:211]: Will keep it here and remove the one inside the sync block ?

Comment (kannanjgithub) [reply opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java:211]: Actually, retain both. The check for `isCallEnded` still serves for performance optimization to avoid monitor acquisition when the call has already finished. The inner check provides the concurrency guard to ensure atomicity between verifying the call lifecycle and creating/mutating the active span.

Comment (AgraVator) [reply opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java:211]: As discussed offline, keeping both

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java (hunk 16 above)
Comment (AgraVator) [reply opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java:236]: As discussed offline, keeping both

**Review by kannanjgithub — APPROVED** [review]

File: opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java [hunk 18]
```diff
@@ -185,25 +200,86 @@ void callEnded(io.grpc.Status status) {
...
+        }
+        recordCallDelayEnd();
+        activeCallDelayType = delayType;
+        activeCallDelaySpan = otelTracer.spanBuilder("Delay")
+            .setParent(Context.current().with(clientSpan))
+            .setAttribute("grpc.delay_type", delayType)
+            .startSpan();
+        activeCallDelaySpan.addEvent(
+            "Delay state transition",
+            Attributes.of(AttributeKey.stringKey("grpc.delay_reason"), delayReason));
```
Comment (AgraVator) [inline opentelemetry/src/main/java/io/grpc/opentelemetry/OpenTelemetryTracingModule.java:230]: Being discussed at https://github.com/grpc/proposal/pull/556#discussion_r3970566534
